apps/prairielearn/assets/scripts/question.ts

```diff
@@ -0,0 +1,162 @@
+  if (!gradingPending) return;
+
+  // By this point, it's safe to open a socket
+  const socket = io('/external-grading');
+
+  socket.emit('init', { variant_id: variantId, variant_token: variantToken }, function (msg) {
+    handleStatusChange(socket, msg);
+  });
+
+  socket.on('change:status', function (msg) {
+    handleStatusChange(socket, msg);
+  });
+}
+
+function handleStatusChange(socket, msg) {
+  msg.submissions.forEach((submission) => {
+    // Always update results
+    updateStatus(submission);
+
+    if (submission.grading_job_status === 'graded') {
+      // Check if this state is reflected in the DOM; it's possible this is
+      // just a message from the initial data sync and that we already have
+      // results in the DOM.
+      const element = document.getElementById('submission-' + submission.id);
+      const status = element.dataset.gradingJobStatus;
+      const gradingJobId = element.dataset.gradingJobId;
+
+      // Ignore jobs that we already have results for, but allow results
+      // from more recent grading jobs to replace the existing ones.
+      if (status !== 'graded' || gradingJobId !== submission.grading_job_id) {
+        // Let's get results for this job!
+        fetchResults(socket, submission.id);
+      }
+    }
+  });
+}
+
+function fetchResults(socket, submissionId) {
+  const {
+    variantId,
+    questionId,
+    instanceQuestionId,
+    variantToken,
+    urlPrefix,
+    questionContext,
+    csrfToken,
+    authorizedEdit,
+  } = (document.querySelector('.question-container') as HTMLElement).dataset;
+
+  const modal = $('#submissionInfoModal-' + submissionId);
+  const wasModalOpen = (modal.data('bs.modal') || {})._isShown;
+  modal.modal('hide');
+
+  socket.emit(
+    'getResults',
+    {
+      question_id: questionId,
+      instance_question_id: instanceQuestionId === '' ? null : instanceQuestionId,
+      variant_id: variantId,
+      variant_token: variantToken,
+      submission_id: submissionId,
+      url_prefix: urlPrefix,
+      question_context: questionContext,
+      csrf_token: csrfToken,
+      // Indicates whether submissions are allowed, either because
+      // the instance question is part of the current user's
+      // assessment instance (authorized_edit==true) or because the
+      // question is open in preview mode (authz_result==undefined)
+      authorized_edit: authorizedEdit,
+    },
+    function (msg) {
+      // We're done with the socket for this incarnation of the page
+      socket.close();
+      if (msg.answerPanel) {
+        const answerContainer = document.querySelector('.answer-body');
+        answerContainer.innerHTML = msg.answerPanel;
+        answerContainer.closest('.grading-block').classList.remove('d-none');
+      }
+      if (msg.submissionPanel) {
+        document.getElementById('submission-' + submissionId).outerHTML = msg.submissionPanel;
+        window.MathJax.typesetPromise();
```

Comment:
I don't think this is safe now that the `async` has been added to the MathJax script tag? We probably want to use `MathJax.config.onReady(() => ...)` to queue up typesetting instead.